server/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -908,7 +908,7 @@ public boolean equals(Object obj) {
         }
 
         IngestDocument other = (IngestDocument) obj;
-        return Objects.equals(ctxMap, other.ctxMap) && Objects.equals(ingestMetadata, other.ingestMetadata);
+        return Objects.equals(ctxMap, other.ctxMap) && Maps.deepEquals(ingestMetadata, other.ingestMetadata);
```

Comment:
I don't think `Maps.deepEquals(...)` happens to work recursively -- if you have string keys and long[] values, it'll do a good job, but if you have string keys and map values (where the maps are string keys to long[] values) then it doesn't work.

I think the consequence for us here is that nested field in an ingest document would still throw this off (e.g. `"somearray"` versus `"nested.somearray"`).

Reply:
Yes I agree. It doesn't work recursively. I will revisit this when I can.

Reply:
https://github.com/elastic/elasticsearch/pull/98901 has been merged which fixes `Maps.deepEquals` to handle this case correctly, and I've added a commit here to make sure we're exercising the nested maps codepath (which would have failed prior to #98901).